crates/nu-protocol/src/engine/stack.rs

```diff
@@ -46,9 +51,67 @@ impl Stack {
             env_hidden: HashMap::new(),
             active_overlays: vec![DEFAULT_OVERLAY_NAME.to_string()],
             recursion_count: 0,
+            parent_stack: None,
+            parent_deletions: vec![],
+        }
+    }
+
+    /// Unwrap a uniquely-owned stack.
+    ///
+    /// In debug mode, this panics if there are multiple references.
+    /// In production this will instead clone the underlying stack.
+    pub fn unwrap_unique(stack_arc: Arc<Stack>) -> Stack {
+        // If you hit an error here, it's likely that you created an extra
+        // Arc pointing to the stack somewhere. Make sure that it gets dropped before
+        // getting here!
+        Arc::try_unwrap(stack_arc).unwrap_or_else(|arc| {
+            // in release mode, we clone the stack, but this can lead to
+            // major performance issues, so we should avoid it
+            debug_assert!(false, "More than one stack reference remaining!");
+            (*arc).clone()
+        })
+    }
+    /// Create a new child stack from a parent.
+    ///
+    /// Changes from this child can be merged back into the parent with
+    /// Stack::with_changes_from_child
+    pub fn with_parent(parent: Arc<Stack>) -> Stack {
+        Stack {
+            // here we are still cloning environment variable-related information
+            env_vars: parent.env_vars.clone(),
+            env_hidden: parent.env_hidden.clone(),
+            active_overlays: parent.active_overlays.clone(),
+            recursion_count: parent.recursion_count,
+            parent_stack: Some(parent),
+            vars: vec![],
+            parent_deletions: vec![],
         }
     }
 
+    /// Take an Arc of a parent (assumed to be unique), and a child, and apply
+    /// all the changes from a child back to the parent.
+    ///
+    /// Here it is assumed that child was created with a call to Stack::with_parent
+    /// with parent
+    pub fn with_changes_from_child(parent: Arc<Stack>, mut child: Stack) -> Stack {
+        // we're going to drop the link to the parent stack on our new stack
+        // so that we can unwrap the Arc as a unique reference
+        //
+        // This makes the new_stack be in a bit of a weird state, so we shouldn't call
+        // any structs
+        drop(child.parent_stack);
+        let mut unique_stack = Stack::unwrap_unique(parent);
+
+        unique_stack.vars.append(&mut child.vars);
+        unique_stack.env_vars = child.env_vars;
+        unique_stack.env_hidden = child.env_hidden;
+        unique_stack.active_overlays = child.active_overlays;
+        unique_stack.recursion_count = child.recursion_count;
```

Comment:
```suggestion
```
In practice, I don't think this will matter, but we should keep the parent's `recursion_count` here to be safe.

Reply:
Yeah, makes sense. I had thought about that problem a bit backwards. Committed this change